File: app/src/test/java/com/amith/animetracker/domain/ImportStatusTest.kt
package com.amith.animetracker.domain

import org.junit.Assert.assertEquals
import org.junit.Test

class ImportStatusTest {

    @Test
    fun `maps each MAL status to the expected imported status`() {
        assertEquals(ImportedEntryStatus.Completed, mapMalListStatus("completed"))
        assertEquals(ImportedEntryStatus.Manual(ManualStatus.CURRENTLY_WATCHING), mapMalListStatus("watching"))
        assertEquals(ImportedEntryStatus.Manual(ManualStatus.CURRENTLY_WATCHING), mapMalListStatus("on_hold"))
        assertEquals(ImportedEntryStatus.Manual(ManualStatus.DROPPED), mapMalListStatus("dropped"))
        assertEquals(ImportedEntryStatus.Manual(ManualStatus.PLAN), mapMalListStatus("plan_to_watch"))
    }

    @Test
    fun `dropped wins over every other status`() {
        val statuses = listOf(
            ImportedEntryStatus.Completed,
            ImportedEntryStatus.Manual(ManualStatus.CURRENTLY_WATCHING),
            ImportedEntryStatus.Manual(ManualStatus.DROPPED),
        )
        assertEquals(ManualStatus.DROPPED, mergeSeriesManualStatus(statuses))
    }

    @Test
    fun `currently watching wins over plan and completed`() {
        val statuses = listOf(
            ImportedEntryStatus.Completed,
            ImportedEntryStatus.Manual(ManualStatus.PLAN),
            ImportedEntryStatus.Manual(ManualStatus.CURRENTLY_WATCHING),
        )
        assertEquals(ManualStatus.CURRENTLY_WATCHING, mergeSeriesManualStatus(statuses))
    }

    @Test
    fun `all plan_to_watch with nothing completed stays PLAN`() {
        val statuses = listOf(
            ImportedEntryStatus.Manual(ManualStatus.PLAN),
            ImportedEntryStatus.Manual(ManualStatus.PLAN),
        )
        assertEquals(ManualStatus.PLAN, mergeSeriesManualStatus(statuses))
    }

    @Test
    fun `completed plus plan_to_watch falls through to auto-derive as NONE`() {
        val statuses = listOf(
            ImportedEntryStatus.Completed,
            ImportedEntryStatus.Manual(ManualStatus.PLAN),
        )
        assertEquals(ManualStatus.NONE, mergeSeriesManualStatus(statuses))
    }

    @Test
    fun `all completed auto-derives as NONE`() {
        val statuses = listOf(ImportedEntryStatus.Completed, ImportedEntryStatus.Completed)
        assertEquals(ManualStatus.NONE, mergeSeriesManualStatus(statuses))
    }

    @Test
    fun `empty list defaults to NONE`() {
        assertEquals(ManualStatus.NONE, mergeSeriesManualStatus(emptyList()))
    }

    @Test
    fun `maps MAL airing status strings, defaulting unknown values to FINISHED`() {
        assertEquals(AiringStatus.AIRING, mapAiringStatus("currently_airing"))
        assertEquals(AiringStatus.NOT_YET_AIRED, mapAiringStatus("not_yet_aired"))
        assertEquals(AiringStatus.FINISHED, mapAiringStatus("finished_airing"))
        assertEquals(AiringStatus.FINISHED, mapAiringStatus(null))
    }
}

File: __tests__/importStatus.test.ts
import { mapMalListStatus, mergeSeriesManualStatus, type ImportedEntryStatus } from '@/domain/importStatus';
import { mapAiringStatus } from '@/domain/reconcileSeries';
import type { ManualStatus } from '@/domain/types';

const COMPLETED: ImportedEntryStatus = { kind: 'COMPLETED' };
const manual = (status: ManualStatus): ImportedEntryStatus => ({ kind: 'MANUAL', status });

test('maps each MAL status to the expected imported status', () => {
  expect(mapMalListStatus('completed')).toEqual(COMPLETED);
  expect(mapMalListStatus('watching')).toEqual(manual('CURRENTLY_WATCHING'));
  expect(mapMalListStatus('on_hold')).toEqual(manual('CURRENTLY_WATCHING'));
  expect(mapMalListStatus('dropped')).toEqual(manual('DROPPED'));
  expect(mapMalListStatus('plan_to_watch')).toEqual(manual('PLAN'));
});

test('dropped wins over every other status', () => {
  const statuses = [COMPLETED, manual('CURRENTLY_WATCHING'), manual('DROPPED')];
  expect(mergeSeriesManualStatus(statuses)).toBe('DROPPED');
});

test('currently watching wins over plan and completed', () => {
  const statuses = [COMPLETED, manual('PLAN'), manual('CURRENTLY_WATCHING')];
  expect(mergeSeriesManualStatus(statuses)).toBe('CURRENTLY_WATCHING');
});

test('all plan_to_watch with nothing completed stays PLAN', () => {
  const statuses = [manual('PLAN'), manual('PLAN')];
  expect(mergeSeriesManualStatus(statuses)).toBe('PLAN');
});

test('completed plus plan_to_watch falls through to auto-derive as NONE', () => {
  const statuses = [COMPLETED, manual('PLAN')];
  expect(mergeSeriesManualStatus(statuses)).toBe('NONE');
});

test('all completed auto-derives as NONE', () => {
  expect(mergeSeriesManualStatus([COMPLETED, COMPLETED])).toBe('NONE');
});

test('empty list defaults to NONE', () => {
  expect(mergeSeriesManualStatus([])).toBe('NONE');
});

test('maps MAL airing status strings, defaulting unknown values to FINISHED', () => {
  expect(mapAiringStatus('currently_airing')).toBe('AIRING');
  expect(mapAiringStatus('not_yet_aired')).toBe('NOT_YET_AIRED');
  expect(mapAiringStatus('finished_airing')).toBe('FINISHED');
  expect(mapAiringStatus(null)).toBe('FINISHED');
});
